application/common/models/Password.php

```diff
@@ -0,0 +1,155 @@
+                    ['minScore' => $this->config['minLength']['value']]
+                ),
+                'when' => function() { return $this->config['minLength']['enabled']; }
+            ],
+            [
+                'password', 'match', 'pattern' => $this->config['maxLength']['phpRegex'],
+                'message' => \Yii::t(
+                    'app',
+                    'Your password exceeds the maximum length of {maxLength}',
+                    ['maxLength' => $this->config['maxLength']['value']]
+                ),
+                'when' => function() { return $this->config['maxLength']['enabled']; }
+            ],
+            [
+                'password', 'match', 'pattern' => $this->config['minNum']['phpRegex'],
+                'message' => \Yii::t(
+                    'app',
+                    'Your password must contain at least {minNum} numbers',
+                    ['minNum' => $this->config['minNum']['value']]
+                ),
+                'when' => function() { return $this->config['minNum']['enabled']; }
+            ],
+            [
+                'password', 'match', 'pattern' => $this->config['minUpper']['phpRegex'],
+                'message' => \Yii::t(
+                    'app',
+                    'Your password must contain at least {minUpper} upper case letters',
+                    ['minUpper' => $this->config['minUpper']['value']]
+                ),
+                'when' => function() { return $this->config['minUpper']['enabled']; }
+            ],
+            [
+                'password', 'match', 'pattern' => $this->config['minSpecial']['phpRegex'],
+                'message' => \Yii::t(
+                    'app',
+                    'Your password must contain at least {minSpecial} special characters',
+                    ['minSpecial' => $this->config['minSpecial']['value']]
+                ),
+                'when' => function() { return $this->config['minSpecial']['enabled']; }
+            ],
+            [
+                'password',
+                ZxcvbnPasswordValidator::className(),
+                'minScore' => $this->config['zxcvbn']['minScore'],
+                'message' => \Yii::t(
+                    'app',
+                    'Your password does not meet the minimum strength of {minScore}',
+                    ['minScore' => $this->config['zxcvbn']['minScore']]
+                ),
+            ],
+        ];
+    }
+
+    /**
+     * Shortcut method to initialize a Password object
+     * @param string $newPassword
+     * @return Password
+     */
+    public static function create($newPassword)
+    {
+        $password = new Password();
+        $password->password = $newPassword;
+
+        return $password;
+    }
+
+    /**
+     * @param int $userId
+     * @param string $username
+     * @return array
+     * @throws BadRequestHttpException
+     * @throws ServerErrorHttpException
+     */
+    public function save($userId, $username)
```

Comment:
This function doesn't seem to do any validation to make sure that the given $userId and $username actually refer to the same user. That seems like something we should verify (if possible).